File: src/commands/submit-claims-from-csv.ts
import * as commander from 'commander';
import { createReadStream, existsSync } from 'fs';
import { parse } from '@fast-csv/parse';
import chalk from 'chalk';

import { actionRunner, serializeError } from '../utils.js';
import { getAccessToken } from '../config.js';
import { createClaim, getBenefitsWithCategories } from '../forma.js';
import { type Claim, claimParamsToCreateClaimOptions } from '../claims.js';
import VERSION from '../version.js';
import { attemptToInferCategoryAndBenefit } from '../openai.js';

const command = new commander.Command();

interface Arguments {
  accessToken?: string;
  inputPath: string;
  openaiApiKey: string;
}

const EXPECTED_HEADERS = [
  'category',
  'benefit',
  'amount',
  'merchant',
  'purchaseDate',
  'description',
  'receiptPath',
];

const readClaimsFromCsv = async (inputPath: string): Promise<Claim[]> => {
  const claims: Claim[] = [];

  return await new Promise((resolve, reject) => {
    createReadStream(inputPath, 'utf8')
      .pipe(parse({ headers: true }))
      .on('error', reject)
      .on('data', (row) => {
        const rowHeaders = Object.keys(row);

        if (
          rowHeaders.length !== EXPECTED_HEADERS.length ||
          !rowHeaders.every((header) => EXPECTED_HEADERS.includes(header))
        ) {
          reject(
            new Error(
              'Invalid CSV headers. Please use a template CSV generated by the `generate-template-csv` command.',
            ),
          );
        }

        const receiptPath = row.receiptPath.split(',').map((path) => path.trim());
        const claim: Claim = { ...row, receiptPath };

        claims.push(claim);
      })
      .on('end', () => {
        resolve(claims);
      });
  });
};

command
  .name('submit-claims-from-csv')
  .version(VERSION)
  .description(
    'Submit multiple Forma claims from a CSV. To generate a template CSV to fill in, use the `generate-template-csv` command. You may attach multiple receipts to a claim by filling the `receipt_path` column with comma-separated paths.',
  )
  .requiredOption('--input-path <input_path>', 'The path to the CSV to read claims from')
  .option('--access-token <access_token>', 'Access token used to authenticate with Forma')
  .option(
    '--openai-api-key <openai_token>',
    'An optional OpenAI API key used to infer the benefit and category based on the merchant and description. If this is set, you may omit the `--benefit` and `--category` options. This can also be configured using the `OPENAI_API_KEY` environment variable.',
    process.env.OPENAI_API_KEY,
  )
  .action(
    actionRunner(async (opts: Arguments) => {
      const { openaiApiKey } = opts;

      const accessToken = opts.accessToken ?? getAccessToken();

      if (!accessToken) {
        throw new Error(
          "You aren't logged in to Forma. Please run `formanator login` first.",
        );
      }

      if (!existsSync(opts.inputPath)) {
        throw new Error(`File '${opts.inputPath}' doesn't exist.`);
      }

      const claims = await readClaimsFromCsv(opts.inputPath);

      if (!claims.length) {
        throw new Error(
          "Your CSV doesn't seem to contain any claims. Have you filled out the template?",
        );
      }

      for (const [index, claim] of claims.entries()) {
        console.log(`Submitting claim ${index + 1}/${claims.length}`);

        try {
          if (claim.benefit !== '' && claim.category !== '') {
            const createClaimOptions = await claimParamsToCreateClaimOptions(
              claim,
              accessToken,
            );
            await createClaim(createClaimOptions);
            console.log(
              chalk.green(`Successfully submitted claim ${index + 1}/${claims.length}`),
            );
          } else if (openaiApiKey) {
            const benefitsWithCategories = await getBenefitsWithCategories(accessToken);
            const { benefit, category } = await attemptToInferCategoryAndBenefit({
              merchant: claim.merchant,
              description: claim.description,
              benefitsWithCategories,
              openaiApiKey,
            });

            const createClaimOptions = await claimParamsToCreateClaimOptions(
              { ...claim, benefit, category },
              accessToken,
            );
            await createClaim(createClaimOptions);
            console.log(
              chalk.green(`Successfully submitted claim ${index + 1}/${claims.length}`),
            );
          } else {
            throw new Error(
              'You must either fill out the `benefit` and `category` columns, or specify an OpenAI API key.',
            );
          }
        } catch (e) {
          console.error(
            chalk.red(
              `Error submitting claim ${index + 1}/${claims.length}: ${serializeError(
                e,
              )}`,
            ),
          );
        }
      }
    }),
  );

export default command;

File: src/commands/submit_claims_from_csv.rs
use anyhow::{Result, bail};
use colored::Colorize;

use crate::claims::{claim_input_to_create_options, read_claims_from_csv};
use crate::cli::SubmitClaimsFromCsvArgs;
use crate::config::resolve_access_token;
use crate::forma::{create_claim, get_benefits_with_categories};
use crate::llm::{infer_all_from_receipt, infer_category_and_benefit};
use crate::verbose;

pub fn run(args: SubmitClaimsFromCsvArgs) -> Result<()> {
    verbose::set(args.verbose);
    let access_token = resolve_access_token(args.access_token.as_deref())?;

    if !args.input_path.exists() {
        bail!("File '{}' doesn't exist.", args.input_path.display());
    }

    let claims = read_claims_from_csv(&args.input_path)?;
    if claims.is_empty() {
        bail!("Your CSV doesn't seem to contain any claims. Have you filled out the template?");
    }

    let benefits = get_benefits_with_categories(&access_token)?;
    let total = claims.len();

    for (index, mut claim) in claims.into_iter().enumerate() {
        let row_number = index + 2;
        println!(
            "Submitting claim {}/{} (row {row_number})",
            index + 1,
            total
        );

        let result = (|| -> Result<()> {
            let has_llm_key = args.openai_api_key.is_some() || args.github_token.is_some();
            let benefit_category_empty = claim.benefit.is_empty() && claim.category.is_empty();
            let other_details_empty = claim.amount.is_empty()
                && claim.merchant.is_empty()
                && claim.purchase_date.is_empty()
                && claim.description.is_empty();
            let other_details_all_filled = !claim.amount.is_empty()
                && !claim.merchant.is_empty()
                && !claim.purchase_date.is_empty()
                && !claim.description.is_empty();

            if !claim.benefit.is_empty() && !claim.category.is_empty() {
                let opts = claim_input_to_create_options(&claim, &access_token)?;
                if args.dry_run {
                    println!("{}", "Dry run: skipping claim submission.".yellow());
                    Ok(())
                } else {
                    create_claim(&opts)
                }
            } else if has_llm_key && benefit_category_empty && other_details_empty {
                if claim.receipt_path.is_empty() {
                    anyhow::bail!(
                        "To infer all claim details from the receipt, you must provide at least one path in the `receiptPath` column."
                    );
                }
                let inferred = infer_all_from_receipt(
                    &claim.receipt_path[0],
                    &benefits,
                    args.openai_api_key.as_deref(),
                    args.github_token.as_deref(),
                )?;
                println!("Inferred amount: {}", inferred.amount);
                println!("Inferred merchant: {}", inferred.merchant);
                println!("Inferred purchase date: {}", inferred.purchase_date);
                println!("Inferred description: {}", inferred.description);
                println!("Inferred benefit: {}", inferred.benefit);
                println!("Inferred category: {}", inferred.category);
                claim.benefit = inferred.benefit;
                claim.category = inferred.category;
                claim.amount = inferred.amount;
                claim.merchant = inferred.merchant;
                claim.purchase_date = inferred.purchase_date;
                claim.description = inferred.description;
                let opts = claim_input_to_create_options(&claim, &access_token)?;
                if args.dry_run {
                    println!("{}", "Dry run: skipping claim submission.".yellow());
                    Ok(())
                } else {
                    create_claim(&opts)
                }
            } else if has_llm_key && benefit_category_empty && other_details_all_filled {
                let inferred = infer_category_and_benefit(
                    &claim.merchant,
                    &claim.description,
                    &benefits,
                    args.openai_api_key.as_deref(),
                    args.github_token.as_deref(),
                )?;
                claim.benefit = inferred.benefit;
                claim.category = inferred.category;
                println!("Inferred benefit: {}", claim.benefit);
                println!("Inferred category: {}", claim.category);
                let opts = claim_input_to_create_options(&claim, &access_token)?;
                if args.dry_run {
                    println!("{}", "Dry run: skipping claim submission.".yellow());
                    Ok(())
                } else {
                    create_claim(&opts)
                }
            } else if has_llm_key {
                anyhow::bail!(
                    "To use LLM inference, a row must either leave every column except `receiptPath` blank (full receipt inference), or fill every column except `benefit` and `category` (benefit/category inference only)."
                );
            } else {
                anyhow::bail!(
                    "You must either fill out the `benefit` and `category` columns, or specify an OpenAI API key or GitHub token."
                );
            }
        })();

        match result {
            Ok(()) => println!(
                "{}",
                format!(
                    "Successfully submitted claim {}/{} (row {row_number})",
                    index + 1,
                    total
                )
                .green()
            ),
            Err(e) => eprintln!(
                "{}",
                format!(
                    "Error submitting claim {}/{}: {e} (row {row_number})",
                    index + 1,
                    total
                )
                .red()
            ),
        }
    }

    Ok(())
}
